prymer/api/picking.py

```diff
@@ -186,285 +79,108 @@ def score(
-        and next_left.length >= left
-        and prev_right.length >= right
-        and next_right.length >= right
-    )
-
-
-def is_acceptable_primer_pair(primer_pair: PrimerPair, params: FilteringParams) -> bool:
-    """Determine if a primer pair can be kept considering:
-
-    1. the primer pair is in the desired size range
-    2. The primer pair has a melting temperature in the desired range
-
-    Args:
-        primer_pair: the primer pair.
-        params: the parameters used for filtering
-
-    Returns:
-       True if the primer pair passes initial checks and can be considered later on.
-    """
-    return (
-        params.amplicon_sizes.min <= primer_pair.amplicon.length <= params.amplicon_sizes.max
-        and params.amplicon_tms.min <= primer_pair.amplicon_tm <= params.amplicon_tms.max
-    )
+    return left_primer.penalty + right_primer.penalty + size_penalty + tm_penalty
 
 
 def build_primer_pairs(
     left_primers: Iterable[Oligo],
     right_primers: Iterable[Oligo],
     target: Span,
-    params: FilteringParams,
+    amplicon_sizes: MinOptMax[int],
+    amplicon_tms: MinOptMax[float],
+    max_heterodimer_tm: Optional[float],
+    weights: PrimerAndAmpliconWeights,
     fasta: FastaFile,
-) -> list[PrimerPair]:
+) -> Iterator[PrimerPair]:
     """Builds primer pairs from individual left and primers.
 
+    Only primer pairs that meet the requirements for amplicon sizes and tms will be returned.
+    In addition, if `max_heterodimer_tm` is provided then primer pairs with a heterodimer tm
+    exceeding the maximum will also be discarded.
+
     Args:
         left_primers: the left primers
         right_primers: the right primers
         target: the genome mapping for the target
-        params: the parameters used for filtering
+        amplicon_sizes: minimum, optimal, and maximum amplicon sizes (lengths)
+        amplicon_tms: minimum, optimal, and maximum amplicon Tms
+        max_heterodimer_tm: if supplied, heterodimer Tms will be calculated for primer pairs,
+          and those exceeding the maximum Tm will be discarded
+        weights: the set of penalty weights
         fasta: the FASTA file from which the amplicon sequence will be retrieved.
 
     Returns:
-        the list of primer pairs, sorted by penalty (increasing)
+        an iterator over all the valid primer pairs, unsorted
     """
-    # generate all the primer pairs
-    primer_pairs: list[PrimerPair] = []
-    for left_primer in left_primers:
-        for right_primer in right_primers:
-            if left_primer.span.refname != right_primer.span.refname:
-                raise ValueError(
-                    "Cannot create a primer pair from left and right primers on different"
-                    f" references: left: '{left_primer.span.refname}'"
-                    f" right: '{right_primer.span.refname}'"
+    # Short circuit if we have no left primers or no right primers
+    if not any(left_primers) or not any(right_primers):
```

Comment:
**issue** If either `left_primers` or `right_primers` is a generator, `any()` will consume an element.

**suggestion** Consider converting the primers to a `list`, or find a solution to peek at the first element without consuming it.

Reply:
Ugh, I always forget the `Iterator` in python is `Iterable`.